### ISSUE TITLE
Translations are not updated when dynamically loading PCK

### ISSUE THREAD

**Issue by mihe** [godotengine/godot#98728]
### Tested versions

Reproducible in 4.4.dev [ef8d981267702de38ffc24136f9d823d31781c60]

### System information

Windows 11 (10.0.22631)

### Issue description

When including updated translation files as part of an exported PCK file that's meant to be dynamically loaded at runtime, you currently have to manually load each `*.translation` file after having loaded the PCK in order for the updated translations to actually have any effect.

The expected behavior would be for the translations to be loaded as part of loading the PCK.

### Steps to reproduce

(The MRP includes a `patch.pck` file that was exported using the patch system introduced in #97118, where the translation with key `HELLO` was changed from "Hello" to "Howdy" before exporting.)

1. Open the MRP.
2. Run the main scene.
3. Note how the label says "Hello".
4. Uncomment the `ResourceLoader.load` line in `main.gd`.
5. Run the main scene again.
6. Note how the label now says "Howdy".

### Minimal reproduction project (MRP)

[localization-patching.zip](https://github.com/user-attachments/files/17600771/localization-patching.zip)

**Comment by PoqXert** [comment]
This happens because localization information is stored in the project settings (project.godot). When pck is loaded dynamically, the project settings are not updated. This issue was mentioned in https://github.com/godotengine/godot-proposals/issues/146#issuecomment-674542744.
[There is a PR](https://github.com/godotengine/godot-proposals/issues/146#issuecomment-674542744) where this issue is solved by loading patches on startup.

**Comment by mihe** [comment]
> [There is a PR](https://github.com/godotengine/godot-proposals/issues/146#issuecomment-674542744) where this issue is solved by loading patches on startup.

@PoqXert Did you intend to link somewhere else? This link was the same as the comment you linked to.

**Comment by PoqXert** [comment]
> Did you intend to link somewhere else? This link was the same as the comment you linked to.

https://github.com/godotengine/godot/pull/65017

**Comment by akien-mga** [comment]
CC @timothyqiu

**Comment by timothyqiu** [comment]
> This happens because localization information is stored in the project settings (project.godot). When pck is loaded dynamically, the project settings are not updated.

In the MRP, the list of translation files is not changed. So it's not related to information stored in project settings.

No resources will be reloaded in `load_resource_pack()`. This behavior is not limited to translation files.

Force reloading the cache for every file loaded from PCK seems to solve the problem. But it's a core behavior change and I'm not sure if it's intended:

```diff
diff --git i/core/io/file_access_pack.cpp w/core/io/file_access_pack.cpp
index 244469e9d9..d10d6e6386 100644
--- i/core/io/file_access_pack.cpp
+++ w/core/io/file_access_pack.cpp
@@ -323,6 +323,14 @@ bool PackedSourcePCK::try_open_pack(const String &p_path, bool p_replace_files,
 			PackedData::get_singleton()->remove_path(path);
 		} else {
 			PackedData::get_singleton()->add_path(p_path, path, file_base + ofs, size, md5, this, p_replace_files, (flags & PACK_FILE_ENCRYPTED), sparse_bundle);
+
+			if (!path.begins_with("res://")) {
+				path = "res://" + path;
+			}
+			Ref<Resource> res = ResourceCache::get_ref(path.trim_suffix(".import"));
+			if (res.is_valid()) {
+				res->reload_from_file();
+			}
 		}
 	}
```

**Comment by mihe** [comment]
Seeing as how we're already doing a bit of magic reloading of things like `uid_cache.bin` and `global_script_class_cache.cfg` when loading a non-main PCK, as seen here:

https://github.com/godotengine/godot/blob/0fdbf050e0c7fc7e0a9d42c2a41ee3bfdffbd8f1/core/config/project_settings.cpp#L573-L579

... means there's already some precedent for reloading this kind of "stuff that's automatically loaded by the engine early in the application lifetime".

With that in mind, maybe there's an argument for doing something similar to what you proposed @timothyqiu but only for `*.translation` files, since they might fall under this category as well?

Having to do it in `PackedSourcePCK` seems very unfortunate though. Maybe it's sufficient to just always reload all the files listed in `internationalization/locale/translations` in `ProjectSettings::_load_resource_pack`, same as what's being done for `uid_cache.bin` and `global_script_class_cache.cfg`?

**Comment by bruvzg** [comment]
> With that in mind, maybe there's an argument for doing something similar to what you proposed @timothyqiu but only for *.translation files, since they might fall under this category as well?

Is it always wanted for all translations? Force-reloading all resources is probably not good, but it might be useful for a lot of them. This might be something that can use an extra file flag in the PCK and be configured when patch PCK is created (not sure what's the best UI for this would be).

```diff
diff --git a/core/io/file_access_pack.cpp b/core/io/file_access_pack.cpp
index 244469e9d9..75c8135119 100644
--- a/core/io/file_access_pack.cpp
+++ b/core/io/file_access_pack.cpp
@@ -305,6 +305,7 @@ bool PackedSourcePCK::try_open_pack(const String &p_path, bool p_replace_files,
                f = fae;
        }

+       Vector<String> resources_to_reload;
        for (int i = 0; i < file_count; i++) {
                uint32_t sl = f->get_32();
                CharString cs;
@@ -323,6 +324,19 @@ bool PackedSourcePCK::try_open_pack(const String &p_path, bool p_replace_files,
                        PackedData::get_singleton()->remove_path(path);
                } else {
                        PackedData::get_singleton()->add_path(p_path, path, file_base + ofs, size, md5, this, p_replace_files, (flags & PACK_FILE_ENCRYPTED), sparse_bundle);
+                       if (flags & PACK_FILE_FORCE_RELOAD) {
+                               if (!path.begins_with("res://")) {
+                                       path = "res://" + path;
+                               }
+                               resources_to_reload.push_back(path);
+                       }
+               }
+       }
+
+       for (const String &path : resources_to_reload) { // After directory processing, to ensure all dependency files are loaded from patch pack.
+               Ref<Resource> res = ResourceCache::get_ref(path.trim_suffix(".import"));
+               if (res.is_valid()) {
+                       res->reload_from_file();
                }
        }
```

**Comment by mihe** [comment]
> Is it always wanted for all translations?

No, ideally you would of course only reload the translations that are actually in the newly loaded PCK. I just figured `PackedSourcePCK::try_open_pack` was the only place where you could reasonably know what files are in the newly loaded PCK, and dealing with this maybe seemed a bit too specific to be in there.

> This might be something that can use an extra file flag in the PCK and be configured when patch PCK is created (not sure what's the best UI for this would be).

I might be straying a bit off-topic here, but this might actually overlap with another patching issue that I've been meaning to look at in the near future, which is the inability to patch already loaded GDScript resources, where things like autoloads (and any of their dependencies), as well as the dependencies of whatever bootstrap script you're using to load the patch PCKs, can't really be patched.

It looks like `Script` does override `reload_from_file` to do some kind of hot-reloading, but `Script::reload` is only called in editor builds. So some shuffling of code might be needed for that to work, but might then benefit from this proposed `PACK_FILE_FORCE_RELOAD` flag as well.

But yeah, I'm not sure what such a UI would look like either, or if it's even useful exposing this to the user. I guess the user would need to first understand that the cost of all the reloading as part of `ProjectSettings.load_resource_pack` is something that can be tweaked, and then find the filtering for it in the export settings, and then somehow understand which files or resource types are the ones taking time, and somehow being aware of the pros/cons of not reloading certain resources. It sounds like a lot.

Maybe going with this `PACK_FILE_FORCE_RELOAD` flag approach, but hardcoding it to only be applied to `*.translation` files (and maybe `*.gd` files if applicable) to begin with, is a good start?

Another problem is that we'd be doing this for every loaded PCK, and in the case of patch PCKs it's not unlikely that projects will be loading multiple PCKs in one go, meaning you'd pay for this reload overhead with each one, as opposed to just once all of them are loaded. That might warrant a new `ProjectSettings.load_resource_packs` method for loading multiple packs in one go or something, but we'd need to defer this proposed reloading in `PackedSourcePCK::try_open_pack` somehow.

**Comment by bruvzg** [comment]
> Maybe going with this PACK_FILE_FORCE_RELOAD flag approach, but hardcoding it to only be applied to *.translation files (and maybe *.gd files if applicable) to begin with, is a good start?

It's probably a good point to start, and if we need more customization, user accessible UI can be always added later.

> That might warrant a new ProjectSettings.load_resource_packs method for loading multiple packs in one go or something, but we'd need to defer this proposed reloading in PackedSourcePCK::try_open_pack somehow.

Makes sense, `try_open_pack` can have a return argument for the resource list/hashset, and actual reloading done in bulk in the `load_resource_pack`/`load_resource_packs` on the whole list after all packs are processed.

**Comment by mihe** [comment]
Great, thank you for the suggestions. I'll have a PR for it up shortly.
